src/backend/cdb/cdbmutate.c

```diff
@@ -1014,6 +1014,12 @@ shareinput_mutator_xslice_2(Node *node, PlannerInfo *root, bool fPop)
 					break;
 			}
 		}
+
+		/*
+		 * Remove writer gang for the shareinput scan readers
+		 */
+		if (currentSlice->gangType == GANGTYPE_PRIMARY_WRITER && !plan->lefttree)
+			currentSlice->gangType = GANGTYPE_PRIMARY_READER;
```

Comment:
@trexxet2, proposed solution raises doubts.
First of all, about this fragment of code. Why do we think that SISC without subplan, when being under writing slice, **always corresponds to the modifying CTE**? I guess, it's possible that GANGTYPE_PRIMARY_WRITER related to other operation, not to the current SISC. And in this case we just overwrite correct `gangType`. 
I manged to find such case:
**Wihtout patch**:
```
set gp_cte_sharing=1;
set optimizer = 0;
create table t_hashed (i int, j int);
create table t_hashed2 (i int, j int);
 explain (slicetable, costs off) with cte1 as (select * from t_hashed)  
update t_hashed2 set j=a.i from cte1 a join (select x.i, x.j from cte1 x limit 1) b on a.i=b.i where b.i=t_hashed2.i;
                                          QUERY PLAN                                          
----------------------------------------------------------------------------------------------
 Update on t_hashed2
   ->  Hash Join
         Hash Cond: (a.i = b.i)
         ->  Subquery Scan on a
               ->  Shared Scan (share slice:id 0:0)
         ->  Hash
               ->  Hash Join
                     Hash Cond: (t_hashed2.i = b.i)
                     ->  Seq Scan on t_hashed2
                     ->  Hash
                           ->  Redistribute Motion 1:3  (slice1; segments: 1)
                                 Hash Key: b.i
                                 ->  Subquery Scan on b
                                       ->  Limit
                                             ->  Gather Motion 3:1  (slice2; segments: 3)
                                                   ->  Limit
                                                         ->  Shared Scan (share slice:id 2:0)
                                                               ->  Seq Scan on t_hashed
 Optimizer: Postgres-based planner
 Slice 0: Primary Writer; root 0; parent -1; gang size 3
 Slice 1: Singleton Reader; root 0; parent 0; gang size 1; segment 1
 Slice 2: Reader; root 0; parent 1; gang size 3
(22 rows)
```
First Shared scan is under the writing slice, however it's a consumer.
**With path**
```
explain (slicetable, costs off) with cte1 as (select * from t_hashed) 
 update t_hashed2 set j=a.i from cte1 a join (select x.i, x.j from cte1 x limit 1) b on a.i=b.i where b.i=t_hashed2.i;
                                          QUERY PLAN                                          
----------------------------------------------------------------------------------------------
 Update on t_hashed2
   ->  Hash Join
         Hash Cond: (a.i = b.i)
         ->  Subquery Scan on a
               ->  Shared Scan (share slice:id 0:0)
         ->  Hash
               ->  Hash Join
                     Hash Cond: (t_hashed2.i = b.i)
                     ->  Seq Scan on t_hashed2
                     ->  Hash
                           ->  Redistribute Motion 1:3  (slice1; segments: 1)
                                 Hash Key: b.i
                                 ->  Subquery Scan on b
                                       ->  Limit
                                             ->  Gather Motion 3:1  (slice2; segments: 3)
                                                   ->  Limit
                                                         ->  Shared Scan (share slice:id 2:0)
                                                               ->  Seq Scan on t_hashed
 Optimizer: Postgres-based planner
 Slice 0: Reader; root 0; parent -1; gang size 3
 Slice 1: Singleton Reader; root 0; parent 0; gang size 1; segment 1
 Slice 2: Primary Writer; root 0; parent 1; gang size 3
(22 rows)
```